situp-api/src/main/java/com/amazon/situp/model/buffer/AbstractBuffer.java

```diff
@@ -0,0 +1,87 @@
+package com.amazon.situp.model.buffer;
+
+import com.amazon.situp.model.configuration.PluginSetting;
+import com.amazon.situp.model.metrics.MetricNames;
+import com.amazon.situp.model.metrics.PluginMetrics;
+import com.amazon.situp.model.record.Record;
+import java.util.Collection;
+import java.util.concurrent.TimeoutException;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.Timer;
+
+/**
+ * Abstract implementation of the Buffer interface to record boilerplate metrics
+ */
+public abstract class AbstractBuffer<T extends Record<?>> implements Buffer<T> {
+    protected final PluginMetrics pluginMetrics;
+    private final Counter recordsWrittenCounter;
+    private final Counter recordsReadCounter;
+    private final Counter writeTimeoutCounter;
+    private final Timer writeTimer;
+    private final Timer readTimer;
+
+    public AbstractBuffer(final PluginSetting pluginSetting) {
+        this.pluginMetrics = PluginMetrics.fromPluginSetting(pluginSetting);
+        this.recordsWrittenCounter = pluginMetrics.counter(MetricNames.RECORDS_WRITTEN);
+        this.recordsReadCounter = pluginMetrics.counter(MetricNames.RECORDS_READ);
+        this.writeTimeoutCounter = pluginMetrics.counter(MetricNames.WRITE_TIMEOUTS);
+        this.writeTimer = pluginMetrics.timer(MetricNames.WRITE_TIME_ELAPSED);
+        this.readTimer = pluginMetrics.timer(MetricNames.READ_TIME_ELAPSED);
+    }
+
+    /**
+     * Records metrics for ingress, time elapsed, and timeouts, while calling the doWrite method
+     * to perform the actual write
+     * @param record the Record to add
+     * @param timeoutInMillis how long to wait before giving up
+     * @throws TimeoutException
+     */
+    @Override
+    public void write(T record, int timeoutInMillis) throws TimeoutException {
+        try {
+            writeTimer.record(() -> {
+                try {
+                    doWrite(record, timeoutInMillis);
+                } catch (TimeoutException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+            recordsWrittenCounter.increment();
+        } catch (RuntimeException e) {
+            if(e.getCause() instanceof TimeoutException) {
+                writeTimeoutCounter.increment();
```

Comment:
Can we not increment the counter in the catch statement above?

Reply:
Sure, that's fine.